app/controllers/admin/users_controller.rb

```diff
@@ -69,6 +78,18 @@ def destroy
       redirect_to admin_user_path(params[:id])
     end
 
+    def export
+      @users = User.select(ATTIBUTES_FOR_CSV + ATTRIBUTES_FOR_LAST_ACTIVITY).includes(:organizations)
+
+      respond_to do |format|
```

Comment:
Currently, we just respond with a downloaded CSV. 

I suspect that for a really large user set like DEV, the server may time out trying to get the records to render. Hence, for the future we may need to evaluate an alternate solution when the server times out, like perhaps creating a background job that will email the csv. 

cc @jgaskins for your thoughts.

Reply:
@Ridhwana YES! Thank you for tagging me on this! 💚 I've dealt with problems with this sort of thing a few times. There are actually multiple problems with the synchronous CSV generation:

- Request timeouts
- Memory bloat

You covered the request timeouts, but another issue is that generating the CSV involves a huge amount of memory because it involves holding 2 complete representations of the user list in memory at the same time — the SQL query results and the CSV. We could probably solve the memory issue since CSV is inherently streamable (using [`ActionController::Live`](https://api.rubyonrails.org/classes/ActionController/Live.html)) and we can chunk the user records up with `find_each`, but that won't fix the request timeout.

Your idea to generate the CSV with a Sidekiq job and email it is very much the solution I've used in the past, with a couple minor twists

- While generating it, we stream it to a tempfile so it doesn't use up all the RAM in the Sidekiq worker
  - This is basically the `ActionController::Live` solution, but without a TCP socket immediately available
  - The Ruby `csv` library supports streaming to an `IO` object so we don't need to use the `ActionController::Renderer`
- The URL we email should not point to the file directly on S3 because we can't un-sign that S3 URL. Instead, we can link to a Forem route, which lets us authorize access to that user dump in case their email is compromised.

I'll still review the current state, but I just wanted to share my thoughts on this localized to where they were requested. 😄

Reply:
This sounds great and exactly along the lines that I was thinking! I may ask some more questions along the way as we get to the state you described. 

Based on the current data sets for Forems (except DEV) do you think that this is an acceptable initial release for the Forems? My thoughts that it is but maybe we can add some monitoring or error handling that will show us when the server times out on this version. Could you provide me with some guidance on that?

Reply:
> Based on the current data sets for Forems (except DEV) do you think that this is an acceptable initial release for the Forems?

Absolutely. For communities up to a few thousand users, doing it synchronously inside the request *should* be feasible. If it's not, I'd be surprised.

> maybe we can add some monitoring or error handling that will show us when the server times out on this version. Could you provide me with some guidance on that?

Absolutely. Would you be up for pairing once this is live?

Reply:
Definitely, I'll DM you and we can schedule some time. Thank you for being so helpful!